### PR TITLE
Fix loop strider reassociation and hoisting logic for off-heap

### PR DESCRIPTION
**Pull request by VermaSh**
To prevent double hoisting of dataAddr pointer, limit reassociation and hoisting criteria to include only those that have dataAddr pointer as their first child.

The offset node doesn't rely on array header size when using dataAddr pointer to get to data elements so no need for reassociation; we can use the original offset node with newly hoisted array load node.

### CODE REVIEW

**Comment by VermaSh** [comment]
For context: `UnsafeSetMemory` test was failing with [1] in off-heap mode. The failure was observed only on x but could have been present on other platforms as well. The failure was due to multiple bugs:
1. Reassociated auto were using header size to get to array data elements instead of loading dataAddr pointer [2].
2. We were erroneously reassociating offset trees [3]. off-heap doesn't require any offset reassociation.
3. Array ref was being double hoisted because there was no check to verify if the array ref is dataAddr pointer or not [4]. This resulted in previously hoisted dataAddr pointer being used as array object during the second pass.

By removing checks for multiply and shift nodes [4], this PR also enables hoisting of dataAddr pointer for unit array stride.

<details open><summary>[1] UnsafeSetMemory failure console snippet</summary>

```
[2024-10-17T06:47:33.075Z] +++ j9vm.test.unsafe.UnsafeSetMemoryTest: +++
[2024-10-17T06:47:33.075Z] command: /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_x86-64_linux_Personal_testList_1/jdkbinary/j2sdk-image/bin/java  -XX:+UseCompressedOops -Xjit -Xgcpolicy:balanced  -Xdump -Xms64m -Xmx64m -classpath /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_x86-64_linux_Personal_testList_1/aqa-tests/TKG/../../jvmtest/functional/VM_Test/VM_Test.jar:/home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_x86-64_linux_Personal_testList_1/../../testDependency/lib/asm-all.jar j9vm.test.unsafe.UnsafeSetMemoryTest 
[2024-10-17T06:47:33.075Z] 
[2024-10-17T06:47:33.075Z] Testing Unsafe.setMemory(Object,long,long,byte) for small arrays.
[2024-10-17T06:47:33.075Z] Testing Unsafe.setMemory(Object,long,long,byte) for large arrays.
[2024-10-17T06:47:34.761Z] Exception in thread "main" java.lang.Error: Found unexpected value 0 at 248233 in set range; expected value 1
[2024-10-17T06:47:34.761Z] 	at j9vm.test.unsafe.UnsafeSetMemoryTest.testSetMemory(UnsafeSetMemoryTest.java:163)
[2024-10-17T06:47:34.761Z] 	at j9vm.test.unsafe.UnsafeSetMemoryTest.testLargeArrays(UnsafeSetMemoryTest.java:102)
[2024-10-17T06:47:34.761Z] 	at j9vm.test.unsafe.UnsafeSetMemoryTest.main(UnsafeSetMemoryTest.java:182)
[2024-10-17T06:47:34.761Z] non-zero exit value: 1
[2024-10-17T06:47:34.761Z] *** Test FAILED *** (j9vm.test.unsafe.UnsafeSetMemoryTest)
```
</details>

[2] [compiler/optimizer/InductionVariable.cpp#L931-L944](https://github.com/eclipse-omr/omr/blob/205c540402c01986f433a34412f9d7fcfe61d90f/compiler/optimizer/InductionVariable.cpp#L931-L944) "Auto reassociation"
[3] [compiler/optimizer/InductionVariable.cpp#L3721-L3725](https://github.com/eclipse-omr/omr/blob/205c540402c01986f433a34412f9d7fcfe61d90f/compiler/optimizer/InductionVariable.cpp#L3721-L3725) "Offset reassocation"
[4] [compiler/optimizer/InductionVariable.cpp#L3657-L3658](https://github.com/eclipse-omr/omr/blob/205c540402c01986f433a34412f9d7fcfe61d90f/compiler/optimizer/InductionVariable.cpp#L3657-L3658) "Off-heap reassociation and hoisting criteria"
[5] [internal issue](https://github.ibm.com/runtimes/openj9-jit-z/issues/915) tracking the failure

**Comment by VermaSh** [comment]
Setting this to WIP for now as I launch personal builds for one last verification.

cc: @r30shah @zl-wang @rmnattas

**Comment by VermaSh** [comment]
`j9vm.test.unsafe.UnsafeSetMemoryTest` passed in my [off-heap personal build](https://hyc-runtimes-jenkins.swg-devops.com/job/Test_openjdk11_j9_sanity.functional_x86-64_linux_Personal_testList_0/2249/consoleText). Waiting for rest of the tests to finish.

**Comment by VermaSh** [comment]
Builds look good, don't see any new failures due to this PR. Marking this ready for review.

**Comment by VermaSh** [comment]
@vijaysun-omr / @hzongaro can I please get a review

File: compiler/optimizer/InductionVariable.cpp [hunk 1]
```diff
@@ -3709,20 +3714,11 @@ bool TR_LoopStrider::reassociateAndHoistComputations(TR::Block *loopInvariantBlo
 
          TR::SymbolReference *internalPointerSymRef = (*_reassociatedAutos)[originalInternalPointerSymbol];
          originalNode->getFirstChild()->recursivelyDecReferenceCount();
-         node->decReferenceCount();
```
Comment (vijaysun-omr) [inline compiler/optimizer/InductionVariable.cpp:3712]: What is the rationale for these changes ? This seems to be unrelated to what was discussed in the commit message ?

Comment (VermaSh) [reply compiler/optimizer/InductionVariable.cpp:3712]: From what I understood, refcount of node (offset node) was being decremented because we were replacing it with it's second child. I think that was just carried over from array access trees with array header size because we don't need to recalculate element offset after hoisting dataAddr node in off-heap mode. I removed the decrement refcount logic as part of my fix for [2nd bug](https://github.com/eclipse-omr/omr/pull/7509#issuecomment-2447431084).

**Comment by vijaysun-omr** [comment]
Jenkins build all

File: compiler/optimizer/InductionVariable.cpp [hunk 2]
```diff
@@ -3709,20 +3714,11 @@ bool TR_LoopStrider::reassociateAndHoistComputations(TR::Block *loopInvariantBlo
...
-         _reassociatedNodes.add(node);
```
Comment (vijaysun-omr) [inline compiler/optimizer/InductionVariable.cpp:3713]: Why do we skip adding to `_reassociatedNodes` ?

Comment (VermaSh) [reply compiler/optimizer/InductionVariable.cpp:3713]: We aren't making any changes to array element offset computation (`1n4`). We simply hoist the dataAddr pointer node (`1n2`) and leave the offset node as is, so we don't need to add anything to `_reassociatedNodes`. Whereas in gencon, we need to update `_reassociatedNodes` because we hoist array header size (`2n5`) from offset node (`2n3`) along with array object load.

```
      off-heap mode:
      1n1   aladd (internal pointer)
      1n2      contiguousArrayDataAddrFieldSymbol (dataAddrPointer, internal pointer)
      1n3         array_obj (pinning array pointer)
      1n4      mul/shift/integer offset

      non off-heap mode:
      2n1   aladd (internal pointer)
      2n2      array_obj (pinning array pointer)
      2n3      add/sub offset
      2n4         index
      2n5         header_size/-header_size
```

**Review by vijaysun-omr — APPROVED** [review]

File: compiler/optimizer/OMRTransformUtil.hpp [hunk 3]
```diff
@@ -226,7 +226,8 @@ class OMR_EXTENSIBLE TransformUtil
    static TR::Node *generateArrayElementAddressTrees(
       TR::Compilation *comp,
       TR::Node *arrayNode,
-      TR::Node *offsetNode = NULL);
+      TR::Node *offsetNode = NULL,
+      TR::Node *originatingByteCodeNode = NULL);
```
Comment (hzongaro) [inline compiler/optimizer/OMRTransformUtil.hpp:230]: It looks like there are existing calls to `generateArrayElementAddressTrees` that will get the default `originatingByteCodeNode` value of NULL, and end up generating `TR::Node`s without the correct bytecode info.

For example, there is a call in [`OMR::TransformUtil::generateFirstArrayElementAddressTrees`](https://github.com/VermaSh/omr/blob/c875d1733e4345cb44744bdbf5a38ab7920a798e/compiler/optimizer/OMRTransformUtil.cpp#L494-L499) which is itself called from [`OMR::ValuePropagation::transformArrayCloneCall`](https://github.com/VermaSh/omr/blob/c875d1733e4345cb44744bdbf5a38ab7920a798e/compiler/optimizer/ValuePropagationCommon.cpp#L4175-L4176).  That transformation is applied for calls to `Object.clone` if the object is known to be an array.

Prior to this change, the IL trees that resulted from that transformation looked like this:

```
n69n        arraycopy  <arraycopy>[#215  helper Method] [flags 0x400 0x0 ] (Unsigned cannotOverflow forwardArrayCopy noArrayStoreCheckArrayCopy wordElementArrayCopy )  [0x7fa62c37f550] bci=[-1,18,5] rc=1 vc=0 vn=- li=- udi=- nc=3 flg=0xd020
n66n          aladd                                                                           [0x7fa62c37f460] bci=[-1,17,5] rc=1 vc=0 vn=- li=- udi=- nc=2
n23n            ==>aload
n65n            lconst 8 (highWordZero X!=0 X>=0 )                                            [0x7fa62c37f410] bci=[-1,0,3] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x4104
n68n          aladd                                                                           [0x7fa62c37f500] bci=[-1,18,5] rc=1 vc=0 vn=- li=- udi=- nc=2
n25n            ==>newarray
n67n            lconst 8 (highWordZero X!=0 X>=0 )                                            [0x7fa62c37f4b0] bci=[-1,0,3] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x4104
n64n          lmul                                                                            [0x7fa62c37f3c0] bci=[-1,18,5] rc=1 vc=0 vn=- li=- udi=- nc=2
n62n            i2l                                                                           [0x7fa62c37f320] bci=[-1,18,5] rc=1 vc=0 vn=- li=- udi=- nc=1
n55n              ==>arraylength
n63n            lconst 4 (highWordZero X!=0 X>=0 )                                            [0x7fa62c37f370] bci=[-1,18,5] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x4104
```

after this change, they look like this:

```
n71n      treetop                                                                             [0x7f52fa57f5f0] bci=[-1,18,5] rc=0 vc=0 vn=- li=- udi=- nc=1
n69n        arraycopy  <arraycopy>[#215  helper Method] [flags 0x400 0x0 ] (Unsigned cannotOverflow forwardArrayCopy noArrayStoreCheckArrayCopy wordElementArrayCopy )  [0x7f52fa57f550] bci=[-1,18,5] rc=1 vc=0 vn=- li=- udi=- nc=3 flg=0xd020
n66n          aladd                                                                           [0x7f52fa57f460] bci=[-1,0,3] rc=1 vc=0 vn=- li=- udi=- nc=2
n23n            ==>aload
n65n            lconst 8 (highWordZero X!=0 X>=0 )                                            [0x7f52fa57f410] bci=[-1,0,3] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x4104
n68n          aladd                                                                           [0x7f52fa57f500] bci=[-1,0,3] rc=1 vc=0 vn=- li=- udi=- nc=2
n25n            ==>newarray
n67n            lconst 8 (highWordZero X!=0 X>=0 )                                            [0x7f52fa57f4b0] bci=[-1,0,3] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x4104
n64n          lmul                                                                            [0x7f52fa57f3c0] bci=[-1,18,5] rc=1 vc=0 vn=- li=- udi=- nc=2
n62n            i2l                                                                           [0x7f52fa57f320] bci=[-1,18,5] rc=1 vc=0 vn=- li=- udi=- nc=1
n55n              ==>arraylength
n63n            lconst 4 (highWordZero X!=0 X>=0 )                                            [0x7f52fa57f370] bci=[-1,18,5] rc=1 vc=0 vn=- li=- udi=- nc=0 flg=0x4104
```

In particular, notice the bytecode info. on the `aladd` nodes.

**Review by hzongaro — COMMENTED** [review]
@VermaSh, sorry for the late review.  I just noticed a slight glitch with the bytecode info. in the IL that's generated that you'll probably want to clean up.

File: compiler/optimizer/OMRTransformUtil.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (VermaSh) [reply compiler/optimizer/OMRTransformUtil.hpp:230]: I have created a [PR](https://github.com/eclipse-omr/omr/pull/7554) to fix one of your suggestions from my last PR. I'll open another PR with your second suggestion.
